libs/vkd3d/device.c

```diff
@@ -3528,29 +3549,80 @@ static D3D12_RESOURCE_BINDING_TIER d3d12_device_determine_resource_binding_tier(
-    caps->options.MaxGPUVirtualAddressBitsPerResource = 40; /* XXX */
-    caps->options.StandardSwizzle64KBSupported = FALSE;
-    caps->options.CrossNodeSharingTier = D3D12_CROSS_NODE_SHARING_TIER_NOT_SUPPORTED;
-    caps->options.CrossAdapterRowMajorTextureSupported = FALSE;
+    options->ConservativeRasterizationTier = D3D12_CONSERVATIVE_RASTERIZATION_TIER_NOT_SUPPORTED;
+    options->MaxGPUVirtualAddressBitsPerResource = 40; /* XXX */
+    options->StandardSwizzle64KBSupported = FALSE;
+    options->CrossNodeSharingTier = D3D12_CROSS_NODE_SHARING_TIER_NOT_SUPPORTED;
+    options->CrossAdapterRowMajorTextureSupported = FALSE;
     /* Requires VK_EXT_shader_viewport_index_layer */
-    caps->options.VPAndRTArrayIndexFromAnyShaderFeedingRasterizerSupportedWithoutGSEmulation = FALSE;
-    caps->options.ResourceHeapTier = D3D12_RESOURCE_HEAP_TIER_2;
+    options->VPAndRTArrayIndexFromAnyShaderFeedingRasterizerSupportedWithoutGSEmulation = FALSE;
+    options->ResourceHeapTier = D3D12_RESOURCE_HEAP_TIER_2;
+}
+
+static void d3d12_device_caps_init_feature_options1(struct d3d12_device *device)
+{
+    D3D12_FEATURE_DATA_D3D12_OPTIONS1 *options1 = &device->d3d12_caps.options1;
+
+    options1->WaveOps = device->d3d12_caps.max_shader_model >= D3D_SHADER_MODEL_6_0;
+
+    if (device->vk_info.EXT_subgroup_size_control)
+    {
+        options1->WaveLaneCountMin = device->device_info.subgroup_size_control_properties.minSubgroupSize;
+        options1->WaveLaneCountMax = device->device_info.subgroup_size_control_properties.maxSubgroupSize;
+    }
+    else
+    {
+        WARN("Device info for WaveLaneCountMin and WaveLaneCountMax may be inaccurate.\n");
+        options1->WaveLaneCountMin = device->device_info.subgroup_properties.subgroupSize;
+        options1->WaveLaneCountMax = device->device_info.subgroup_properties.subgroupSize;
+    }
+
+    if (device->vk_info.AMD_shader_core_properties)
+    {
+        const VkPhysicalDeviceShaderCorePropertiesAMD *amd = &device->device_info.shader_core_properties;
+        const VkPhysicalDeviceShaderCoreProperties2AMD *amd2 = &device->device_info.shader_core_properties2;
+        uint32_t compute_units;
+        
+        if (device->vk_info.AMD_shader_core_properties2)
+            compute_units = amd2->activeComputeUnitCount;
+        else
+            compute_units = amd->shaderEngineCount *
+                    amd->computeUnitsPerShaderArray *
+                    amd->wavefrontSize;
+
+        /* There doesn't seem to be a way to get the SIMD width,
+         * so just assume 64 lanes per CU. True on GCN and RDNA. */
+        options1->TotalLaneCount = compute_units * 64;
+    }
+    else if (device->vk_info.NV_shader_sm_builtins)
+    {
+        const VkPhysicalDeviceShaderSMBuiltinsPropertiesNV *nv = &device->device_info.shader_sm_builtins_properties;
+        options1->TotalLaneCount = nv->shaderSMCount * nv->shaderWarpsPerSM;
```

Comment:
This needs to be multiplied by 32 I think. This is the number of warps, not threads.

Reply:
On my 1660 Ti I get 24 for SM count and 32 for Warp count.

Reply:
Hm, looking at more GPUInfo reports, it actually looks like we have a similar problem as with AMD in that this only tells us the number of warps in flight, not the number of threads per SM. On Pascal/Turing it looks like we want to multiply by 2, on Volta we don't want to multiply by anything, haven't checked Maxwell or Kepler. But yeah, this is inaccurate.

Reply:
In that case we should not look at shaderWarpsPerSM, as that's occupancy, the more natural formulation would be something like `SMCount * subgroupSize * magic 2 factor for Turing`.